diff --git a/.gitignore b/.gitignore
index 8f53d915b4..795dcc2de5 100644
--- a/.gitignore
+++ b/.gitignore
@@ -23,3 +23,7 @@ nbproject/*
 cmake-build-*/
 .vs
 *.user
+
+#BotEngine
+/repack/*
+/src/scripts/Custom/BotEngine/*
\ No newline at end of file
diff --git a/src/server/game/Handlers/CharacterHandler.cpp b/src/server/game/Handlers/CharacterHandler.cpp
index 5f8a4abd36..4f5f663a65 100644
--- a/src/server/game/Handlers/CharacterHandler.cpp
+++ b/src/server/game/Handlers/CharacterHandler.cpp
@@ -697,6 +697,31 @@ void WorldSession::HandleCharDeleteOpcode(WorldPacket& recvData)
     SendCharDelete(CHAR_DELETE_SUCCESS);
 }
 
+/*>>>BotEngine*/
+void WorldSession::HandleBotLogin(ObjectGuid pmCharacterGUID)
+{
+    if (PlayerLoading() || GetPlayer() != nullptr)
+    {
+        TC_LOG_ERROR("network", "Player tries to login again, AccountId = %d", GetAccountId());
+        KickPlayer("WorldSession::HandlePlayerLoginOpcode Another client logging in");
+        return;
+    }
+
+    m_playerLoading = true;
+    std::shared_ptr<LoginQueryHolder> holder = std::make_shared<LoginQueryHolder>(GetAccountId(), pmCharacterGUID);
+
+    if (!holder->Initialize())
+    {
+        m_playerLoading = false;
+        return;
+    }
+
+    AddQueryHolderCallback(CharacterDatabase.DelayQueryHolder(holder)).AfterComplete([this](SQLQueryHolderBase const& holder) {
+        HandlePlayerLogin(static_cast<LoginQueryHolder const&>(holder));
+    });
+}
+/*<<<BotEngine*/
+
 void WorldSession::HandlePlayerLoginOpcode(WorldPacket& recvData)
 {
     if (PlayerLoading() || GetPlayer() != nullptr)
@@ -1006,6 +1031,16 @@ void WorldSession::HandlePlayerLogin(LoginQueryHolder const& holder)
     sScriptMgr->OnPlayerLogin(pCurrChar, firstLogin);
 
     TC_METRIC_EVENT("player_events", "Login", pCurrChar->GetName());
+
+    /*>>>BotEngine*/
+    if (_isBotSession)
+    {
+        if (!sCharacterCache->HasCharacterCacheEntry(pCurrChar->GetGUID()))
+        {
+            sCharacterCache->AddCharacterCacheEntry(pCurrChar->GetGUID(), GetAccountId(), pCurrChar->GetName(), pCurrChar->GetGender(), pCurrChar->GetRace(), pCurrChar->GetClass(), pCurrChar->GetLevel());
+        }
+    }
+    /*<<<BotEngine*/
 }
 
 void WorldSession::SendFeatureSystemStatus()
diff --git a/src/server/game/Server/WorldSession.cpp b/src/server/game/Server/WorldSession.cpp
index 6a1713c8fe..ba21ced29c 100644
--- a/src/server/game/Server/WorldSession.cpp
+++ b/src/server/game/Server/WorldSession.cpp
@@ -206,6 +206,11 @@ void WorldSession::SendPacket(WorldPacket const* packet)
 {
     ASSERT(packet->GetOpcode() != NULL_OPCODE);
 
+    /*>>>BotEngine*/
+    if (_isBotSession)
+        return;
+    /*<<<BotEngine*/
+
     if (!m_Socket)
         return;
 
@@ -287,6 +292,29 @@ void WorldSession::LogUnprocessedTail(WorldPacket* packet)
 /// Update the WorldSession (triggered by World update)
 bool WorldSession::Update(uint32 diff, PacketFilter& updater)
 {
+    /*>>>BotEngine*/
+    if (_isBotSession) {
+        ProcessQueryCallbacks();
+
+        if (!_player)
+            return true;
+
+        if (_player->IsBeingTeleportedNear())
+        {
+            WorldPacket data(MSG_MOVE_TELEPORT_ACK, 10);
+            data << _player->GetGUID().WriteAsPacked();
+            data << uint32(0) << uint32(0);
+            HandleMoveTeleportAck(data);
+        }
+        else if (_player->IsBeingTeleportedFar())
+        {
+            HandleMoveWorldportAck();
+        }
+
+        return true;
+    }
+    /*<<<BotEngine*/
+
     ///- Before we process anything:
     /// If necessary, kick the player because the client didn't send anything for too long
     /// (or they've been idling in character select)
diff --git a/src/server/game/Server/WorldSession.h b/src/server/game/Server/WorldSession.h
index 165217e939..aedc65665c 100644
--- a/src/server/game/Server/WorldSession.h
+++ b/src/server/game/Server/WorldSession.h
@@ -667,6 +667,12 @@ class TC_GAME_API WorldSession
         time_t GetCalendarEventCreationCooldown() const { return _calendarEventCreationCooldown; }
         void SetCalendarEventCreationCooldown(time_t cooldown) { _calendarEventCreationCooldown = cooldown; }
 
+        /*>>>BotEngine*/
+        bool IsBot() const { return _isBotSession; }
+        void IsBot(bool flag) { _isBotSession = flag; }
+        void HandleBotLogin(ObjectGuid pmCharacterGUID);
+        /*<<<BotEngine*/
+
     public:                                                 // opcodes handlers
 
         void Handle_NULL(WorldPacket& recvPacket);          // not used
@@ -1329,6 +1335,10 @@ class TC_GAME_API WorldSession
 
         WorldSession(WorldSession const& right) = delete;
         WorldSession& operator=(WorldSession const& right) = delete;
+
+        /*>>>BotEngine*/
+        bool _isBotSession = false;
+        /*<<<BotEngine*/
 };
 #endif
 /// @}
diff --git a/src/server/game/World/World.cpp b/src/server/game/World/World.cpp
index da1e9654cd..8007c1358b 100644
--- a/src/server/game/World/World.cpp
+++ b/src/server/game/World/World.cpp
@@ -2646,6 +2646,10 @@ void World::Update(uint32 diff)
         sMetric->Update();
         TC_METRIC_VALUE("update_time_diff", diff);
     }
+
+    /*>>>BotEngine*/
+    CallPostUpdateCallbacks(diff);
+    /*<<<BotEngine*/
 }
 
 void World::ForceGameEventUpdate()
diff --git a/src/server/game/World/World.h b/src/server/game/World/World.h
index 109b81f7db..5ff9a6e1cf 100644
--- a/src/server/game/World/World.h
+++ b/src/server/game/World/World.h
@@ -780,6 +780,28 @@ class TC_GAME_API World
 #ifdef ELUNA
         Eluna* GetEluna() const { return sElunaMgr->Get(_elunaInfo); }
 #endif
+
+        /*>>>BotEngine*/
+        template<class T> void AddPostUpdateCallback(T* const obj, void(T::* const func)(uint32), bool reserve = false) {
+            if (!reserve)
+                _postUpdateCallbacks.emplace_back(std::bind(func, obj, std::placeholders::_1));
+            else
+                _postUpdateCallbacksReserved.emplace_back(std::bind(func, obj, std::placeholders::_1));
+        };
+        void CallPostUpdateCallbacks(uint32 diff)
+        {
+            for (const auto& cb : _postUpdateCallbacks)
+                cb(diff);
+            _postUpdateCallbacks.clear();
+
+            if (_postUpdateCallbacksReserved.size()) {
+                // deep copy reserved callbacks
+                _postUpdateCallbacks.assign(_postUpdateCallbacksReserved.begin(), _postUpdateCallbacksReserved.end());
+                _postUpdateCallbacksReserved.clear();
+            }
+        }
+        /*<<<BotEngine*/
+
     protected:
         void _UpdateGameTime();
 
@@ -893,6 +915,12 @@ class TC_GAME_API World
         bool _guidAlert;
         uint32 _warnDiff;
         time_t _warnShutdownTime;
+
+        /*>>>BotEngine*/
+        std::vector<std::function<void(int)>> _postUpdateCallbacks;
+        std::vector<std::function<void(int)>> _postUpdateCallbacksReserved;
+        /*<<<BotEngine*/
+
 #ifdef ELUNA
         ElunaInfo _elunaInfo;
 #endif
